lib/shopping_cart.rb

```diff
@@ -0,0 +1,49 @@
+class ShoppingCart
+  attr_reader :name, :capacity, :products, :details, :is_full
+
+  def initialize(name, capacity)
+    @name = name
+    @capacity = capacity.delete_suffix("items").to_i
+    @products = []
+    @details = {}
+    @details[:name] = name
+    @details[:capacity] = @capacity
+    @is_full = false
+
+  end
+
+  def add_product(product)
+    @products << product
+    total_products = products.sum do |product|
+      product.quantity
+    end
+    @is_full = true if total_products >= capacity
+  end
+
+  def is_full?
+    is_full
+  end
+
+  def products_by_category(category)
+    @products.select {|product| product.category == category}
+  end
+
+  def percentage_occupied
+    total_products = products.sum do |product|
+      product.quantity
+    end
```

Comment:
You probably noticed that you're repeating this code, this is also a great reason why pulling this into it's own method would be beneficial.